File: antidescuadre/src/logica/cuentas.ts
import { db } from '../db/db'
import type { Cuenta, ItemCuenta, Pago, SeleccionVariante, Producto } from '../db/tipos'
import { redondear } from './dinero'

export function totalItem(item: ItemCuenta): number {
  return redondear(item.precioUnitario * item.cantidad)
}

export function totalItems(items: ItemCuenta[]): number {
  return redondear(items.reduce((s, it) => s + totalItem(it), 0))
}

export function totalPagado(pagos: Pago[]): number {
  return redondear(pagos.reduce((s, p) => s + p.monto, 0))
}

export function saldoCuenta(items: ItemCuenta[], pagos: Pago[]): number {
  return redondear(totalItems(items) - totalPagado(pagos))
}

export function precioUnitario(producto: Producto, variantes: SeleccionVariante[]): number {
  return redondear(producto.precio + variantes.reduce((s, v) => s + v.delta, 0))
}

export function etiquetaVariantes(variantes: SeleccionVariante[]): string {
  return variantes.map(v => v.camino.join(' · ')).join(' · ')
}

// ---------- Operaciones sobre cuentas ----------

export async function abrirCuentaEnMesa(mesaId: number, turnoId: number): Promise<number> {
  return db.cuentas.add({
    mesaId, turnoId, estado: 'abierta', abiertaEn: Date.now(), esDirecta: false,
  })
}

export async function cuentaActivaDeMesa(mesaId: number): Promise<Cuenta | undefined> {
  return db.cuentas.where({ mesaId }).filter(c => c.estado === 'abierta').first()
}

export interface LineaNueva {
  producto: Producto
  variantes: SeleccionVariante[]
  cantidad: number
}

// Agrega una tanda de productos a una cuenta. Los ítems nacen "pendientes"
// (regla 12); en venta directa nacen entregados porque se sirven en barra.
export async function agregarTanda(
  cuentaId: number, lineas: LineaNueva[], entregadoInmediato = false,
): Promise<void> {
  const ahora = Date.now()
  await db.items.bulkAdd(lineas.map(l => ({
    cuentaId,
    productoId: l.producto.id ?? null,
    nombre: l.producto.nombre,
    variantes: l.variantes,
    cantidad: l.cantidad,
    precioUnitario: precioUnitario(l.producto, l.variantes),
    estado: entregadoInmediato ? 'entregado' as const : 'pendiente' as const,
    agregadoEn: ahora,
    entregadoEn: entregadoInmediato ? ahora : undefined,
    tandaId: ahora,
  })))
}

export async function cerrarCuenta(cuentaId: number): Promise<void> {
  await db.cuentas.update(cuentaId, { estado: 'cerrada', cerradaEn: Date.now() })
}

// ---------- Turno ----------

// Devuelve null (no undefined) cuando no hay turno, para que useLiveQuery
// pueda distinguir "cargando" (undefined) de "sin turno activo" (null).
export async function turnoActivo() {
  return (await db.turnos.where('estado').equals('activo').first()) ?? null
}

export async function iniciarTurno(): Promise<number> {
  return db.turnos.add({ inicio: Date.now(), estado: 'activo' })
}

export interface MesaConSaldo { mesaId: number | null; alias: string; saldo: number }

// Mesas del turno con saldo pendiente (bloquean el cierre — regla 1b).
export async function mesasConSaldoPendiente(turnoId: number): Promise<MesaConSaldo[]> {
  const abiertas = await db.cuentas.where({ turnoId }).filter(c => c.estado === 'abierta').toArray()
  const resultado: MesaConSaldo[] = []
  for (const c of abiertas) {
    const items = await db.items.where({ cuentaId: c.id! }).toArray()
    const pagos = await db.pagos.where({ cuentaId: c.id! }).toArray()
    const saldo = saldoCuenta(items, pagos)
    if (saldo > 0) {
      const mesa = c.mesaId != null ? await db.mesas.get(c.mesaId) : undefined
      resultado.push({ mesaId: c.mesaId, alias: mesa?.alias ?? 'Venta directa', saldo })
    }
  }
  return resultado
}

// Cierra el turno si ninguna mesa tiene saldo; las cuentas abiertas en $0 se cierran.
export async function cerrarTurno(turnoId: number): Promise<{ ok: boolean; bloqueos: MesaConSaldo[] }> {
  const bloqueos = await mesasConSaldoPendiente(turnoId)
  if (bloqueos.length > 0) return { ok: false, bloqueos }
  const abiertas = await db.cuentas.where({ turnoId }).filter(c => c.estado === 'abierta').toArray()
  for (const c of abiertas) await cerrarCuenta(c.id!)
  await db.turnos.update(turnoId, { estado: 'cerrado', fin: Date.now() })
  return { ok: true, bloqueos: [] }
}

File: app/lib/logica/cuentas.dart
// Lógica pura de cuentas: totales, saldos, precios con variantes y las
// operaciones de negocio (tandas, cierre de cuenta y de turno con su
// restricción — reglas 1b, 5, 6 y 12 del documento de planificación).

import 'package:drift/drift.dart';

import '../datos/base.dart';
import '../datos/modelos.dart';
import 'dinero.dart';

double totalItem(Item item) => redondear(item.precioUnitario * item.cantidad);

double totalItems(List<Item> lista) =>
    redondear(lista.fold(0.0, (s, it) => s + totalItem(it)));

double totalPagado(List<Pago> lista) =>
    redondear(lista.fold(0.0, (s, p) => s + p.monto));

double saldoCuenta(List<Item> itemsCuenta, List<Pago> pagosCuenta) =>
    redondear(totalItems(itemsCuenta) - totalPagado(pagosCuenta));

double precioUnitario(Producto producto, List<SeleccionVariante> variantes) =>
    redondear(producto.precio + variantes.fold(0.0, (s, v) => s + v.delta));

// Reparte un saldo en n partes iguales a centavos; la última absorbe el residuo.
List<double> partesIguales(double saldo, int n) {
  final base = (saldo / n * 100).floorToDouble() / 100;
  return List.generate(
    n,
    (i) => i == n - 1 ? redondear(saldo - base * (n - 1)) : base,
  );
}

// Línea elegida en el selector, aún no confirmada.
class LineaNueva {
  final Producto producto;
  final List<SeleccionVariante> variantes;
  final int cantidad;

  const LineaNueva({required this.producto, required this.variantes, required this.cantidad});

  double get unitario => precioUnitario(producto, variantes);
  double get total => redondear(unitario * cantidad);
}

// Agrega una tanda de productos a una cuenta. Los ítems siempre nacen
// "pendientes" (regla 12).
Future<void> agregarTanda(
  BaseDatos db,
  int cuentaId,
  List<LineaNueva> lineas,
) async {
  final ahora = DateTime.now();
  await db.batch((b) {
    b.insertAll(db.items, [
      for (final l in lineas)
        ItemsCompanion.insert(
          cuentaId: cuentaId,
          productoId: Value(l.producto.id),
          nombre: l.producto.nombre,
          variantesJson: Value(seleccionesAJson(l.variantes)),
          cantidad: l.cantidad,
          precioUnitario: l.unitario,
          estado: const Value('pendiente'),
          agregadoEn: ahora,
          tandaId: ahora.millisecondsSinceEpoch,
        ),
    ]);
  });
}

class MesaConSaldo {
  final int? mesaId;
  final String alias;
  final double saldo;
  const MesaConSaldo({required this.mesaId, required this.alias, required this.saldo});
}

// Mesas del turno con saldo pendiente: bloquean el cierre (regla 1b).
Future<List<MesaConSaldo>> mesasConSaldoPendiente(BaseDatos db, int turnoId) async {
  final abiertas = await (db.select(db.cuentas)
        ..where((c) => c.turnoId.equals(turnoId) & c.estado.equals('abierta')))
      .get();
  final resultado = <MesaConSaldo>[];
  for (final c in abiertas) {
    final itemsCuenta =
        await (db.select(db.items)..where((i) => i.cuentaId.equals(c.id))).get();
    final pagosCuenta =
        await (db.select(db.pagos)..where((p) => p.cuentaId.equals(c.id))).get();
    final saldo = saldoCuenta(itemsCuenta, pagosCuenta);
    if (saldo > 0) {
      final mesa = c.mesaId == null
          ? null
          : await (db.select(db.mesas)..where((m) => m.id.equals(c.mesaId!)))
              .getSingleOrNull();
      resultado.add(MesaConSaldo(
        mesaId: c.mesaId,
        alias: mesa?.alias ?? 'Venta directa',
        saldo: saldo,
      ));
    }
  }
  return resultado;
}

class ResultadoCierre {
  final bool ok;
  final List<MesaConSaldo> bloqueos;
  const ResultadoCierre(this.ok, this.bloqueos);
}

// Cierra el turno solo si ninguna mesa tiene saldo; las cuentas abiertas
// en $0 se cierran junto con el turno.
Future<ResultadoCierre> cerrarTurno(BaseDatos db, int turnoId) async {
  final bloqueos = await mesasConSaldoPendiente(db, turnoId);
  if (bloqueos.isNotEmpty) return ResultadoCierre(false, bloqueos);
  final abiertas = await (db.select(db.cuentas)
        ..where((c) => c.turnoId.equals(turnoId) & c.estado.equals('abierta')))
      .get();
  for (final c in abiertas) {
    await db.cerrarCuenta(c.id);
  }
  await (db.update(db.turnos)..where((t) => t.id.equals(turnoId))).write(TurnosCompanion(
    estado: const Value('cerrado'),
    fin: Value(DateTime.now()),
  ));
  return const ResultadoCierre(true, []);
}
